[[ Bug 14316 ]] Allow infinite result in div, times and pow operations

livecodeali opened the pull request:
Before merging, we need to check the siginfo struct upon handling the SIGFPE signal to set `errno` more specifically, as otherwise we are conflating eg divide by zero and out of range errors.

livecodeali committed:
[[ Bug 14316 ]] Allow infinite result in div, times and pow operations

livecodeali committed:
[[ Math ]] Standardardize domain and range error checking

This patch templatizes the math functions in order to standardize
the checks for domain and range errors in their computation. Any 
non-finite result from finite inputs is deemed an error - we use
floating point exception checking to distinguish domain errors 
(FE_INVALID and FE_DIVIDE_BY_ZERO) and range errors (FE_UNDERFLOW and
FE_OVERFLOW).

In many cases this was not previously being checked, hence a number
of new errors have been added to deal with the errors that may occur
for each math operation.

This patch also moves MCU_dofunc from utils to the exec-math module, 
and where it makes sense, splits off certain functions into their
own implementations. This prevents one value having to be recomputed
in each iteration of the values passed to `geometricMean`, as well as
removing the computation of the arithmetic mean for several functions
that don't use it.

Finally, this patch removes unnecessary division operations from the 
wrap and mod implementations. Although fmod does not appear to raise
the FE_INVALID exception when its second argument is zero, the result
is still non-finite and thus throw the correct error by defaulting to
the domain error in MCMathThrowFloatingPointException.

livecodeali committed:
[[ Math ]] Add tests for stats functions, errors and infinity handling

This patch adds three new sets of tests:          
Firstly it adds tests for the stats functions which were previously    
untested. Secondly it adds tests for all the possible domain and range
errors that can be caused by math function computations, and finally   
it adds tests for all types of computations involving infinity which 
neither throw an error nor return NaN.

runrevmark commented on engine/src/executionerrors.h:
```diff
@@ -2778,8 +2778,16 @@ enum Exec_errors
+    EE_PROPERTY_NOTADATA,
     
+    // {EE-0911} Math: domain error
+    EE_MATH_DOMAIN,
```
We are probably safe to re-use the first DOMAIN / RANGE / DIV BY ZERO code that exists, and just mark all others as unused... So EE_ACOS_DOMAIN -> EE_MATH_DOMAIN (numeric: domain error (undefined)); EE_DIVIDE_RANGE -> EE_MATH_RANGE (numeric: range error (overflow)) ; EE_DIVIDE_ZERO -> EE_MATH_ZERO (numeric: divide by zero).

runrevmark replied:
Codes which they replace should be changed to EE_UNUSED_XXXX - where XXXX is the error code (formatted with 0s).

runrevmark commented:
@livecodeali Apart from the suggestion about re-using three existing error codes - this looks absolutely great.

runrevmark commented on docs/notes/bugfix-14316.md:
```diff
@@ -0,0 +1 @@
+# Don't throw error when infinity used innocuously in arithmetic expressions
```
The release note should probably describe the changes - i.e. when domain/range/div-by-zero are thrown and the unification of the error codes.

livecodeali committed:
[[ Math ]] Clarify new math op error behavior in documentation

This patch adds a reference to a new math operation glossary entry 
for each math operation in the dictionary, in which the rules for
determining what execution error is thrown (if any) are described.

livecodeali committed:
[[ Math ]] Add release note describing new standard math error checking

This patch adds a release note describing how math operations are now
checked for errors.

runrevmark commented:
@livecode-vulcan review ok 2027302

livecode-vulcan commented:
💙 review by @runrevmark ok 20273023ac01c8dd223bc4d85e851371a38053f8

livecode-vulcan commented:
😎 test success 20273023ac01c8dd223bc4d85e851371a38053f8

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/479)
* try-community-armv7-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-sdk26_ndk16r15/builds/471)
* try-community-arm64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-sdk26_ndk16r15/builds/471)
* try-community-x86-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-sdk26_ndk16r15/builds/471)
* try-community-x86_64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-sdk26_ndk16r15/builds/471)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/1993)
* try-community-universal-ios-iphoneos12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos12.1/builds/221)
* try-community-universal-ios-iphonesimulator12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator12.1/builds/220)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/2024)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/2015)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/2013)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/2047)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/1984)
